agent_build_refactored/scripts/get_github_actions_job_matrices.py

```diff
@@ -0,0 +1,189 @@
+    version_file_path = SOURCE_ROOT / "VERSION"
+    current_version = version_file_path.read_text().strip()
+    if GITHUB_REF_NAME == f"v{current_version}":
+        limited_run = False
+
+
+ALL_USED_RUNNERS = {
+    **ALL_IMAGE_BUILDERS
+}
+
+
+def _get_runners_all_pre_built_steps(runners: List[Type[Runner]]) -> Dict[str, RunnerStep]:
+    """
+    Get collection of all RunnerSteps that are used by the Runners from the given list.
+    :return: Dict with all used RunnerSteps. Step ID - key, step - value.
+    """
+    _all_runner_steps = {}
+    for runner_cls in runners:
+        for step in runner_cls.get_all_cacheable_steps():
+            if not step.github_actions_settings.pre_build_in_separate_job:
+                continue
+            _all_runner_steps[step.id] = step
+    return _all_runner_steps
+
+
+# Search for all steps that are used by given runners.
+_all_used_runner_steps = _get_runners_all_pre_built_steps(
+    runners=list(ALL_USED_RUNNERS.values())
+)
+
+# Create runner for each found pre-built step.
+_pre_built_step_runners = {}
+for step in _all_used_runner_steps.values():
+    # Create "dummy" Runner for each runner step that has to be pre-built, this dummy runner will be executed
+    # by its fqdn to run the step.
+    class StepWrapperRunner(Runner):
+        REQUIRED_STEPS = [step]
+
+    # Since this runner class is created dynamically we have to generate a constant fqdn for it.
+    StepWrapperRunner.assign_fully_qualified_name(
+        class_name="pre-built-",
+        module_name=__name__,
+        class_name_suffix=step.id,
+    )
+    _pre_built_step_runners[step.id] = StepWrapperRunner
+
+
+def main():
+    run_type_name = "limited" if limited_run else "full"
+    print(
+        f"Doing {run_type_name} workflow run.",
+        file=sys.stderr
+    )
+    print(
+        f"event_name: {GITHUB_EVENT_NAME}\n"
+        f"base_ref: {GITHUB_BASE_REF}\n"
+        f"ref_type: {GITHUB_REF_TYPE}\n"
+        f"ref_name: {GITHUB_REF_NAME}",
+        file=sys.stderr,
+    )
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--images-build-matrix-json-file",
+        dest="images_build_matrix_json_file",
+        required=True,
+        help="Path to a JSON file with images build job matrix."
+    )
+
+    args = parser.parse_args()
+    images_build_matrix_json_file_path = pl.Path(args.images_build_matrix_json_file)
+
+    images_build_matrix = json.loads(
+        images_build_matrix_json_file_path.read_text()
+    )
+
+    # List of all runners that are used by this workflow run.
+    used_runners = []
+
+    # Generate a final agent image build job matrix and filter out job for non-limited run, if needed.
+    result_images_build_matrix = {"include": []}
+    for job in images_build_matrix:
+        is_basic = job.get("basic", False)
+
+        # Skip job which id not basic if run is not limited.
+        if not is_basic and limited_run:
```

Comment:
I think the comment is a bit confusing and should probably say "Skip job with basic attribute for limited run".